Change the image server to Amazon S3 from public folder of the app

tokuchan-pg opened the pull request:
# WHAT
メッセージ送信機能で画像が送信された際の画像の保存先を、アプリケーションからS3に変更する

# WHY
画像の保存先がアプリケーション内だと容量を圧迫してしまうため

### 画像送信完了後のページのソース
![2017-05-04 15 08 30](https://cloud.githubusercontent.com/assets/25572309/25692244/144d3bb8-30dc-11e7-8e72-0ec43863c189.png)
 
### 画像送信完了後のS3バケット画面
![2017-05-04 15 13 04](https://cloud.githubusercontent.com/assets/25572309/25692279/3e9221d6-30dc-11e7-967a-ea1b67f2a4a5.png)

tokuchan-pg committed:
install gem 'fog' & 'dotenv-rails'

tokuchan-pg committed:
revised the uploader setting

tokuchan-pg committed:
generated carrierwave.rb to save images in S3

made git ignore .env file in which S3 KEY is written.

vip-take commented:
カリキュラムにも書いてありますが、環境に応じて、画像の保存先が変更されるようにしましょう！
```
【development,production環境】 fog(s3)
【test環境】 file(local)
```

vip-take commented:
いつもわかりやすいプルリクですね！

tokuchan-pg committed:
swich the image server on rails environment (development or test)

tokuchan-pg commented:
>カリキュラムにも書いてありますが、環境に応じて、画像の保存先が変更されるようにしましょう！
```
【development,production環境】 fog(s3)
【test環境】 file(local)
```

修正いたしました！

【ご質問】
ご指摘の通りcarrierwave.rbの中で環境に応じて画像の保存先が変わるように設定しましたが、この場合、image_uploder.rbの中のstorageの指定の記述はcarrierwave.rbの記述よりも優先度が下がるという形になりますでしょうか？（carrierwave.rbの中で設定していればimage_uploder.rbでの指定は不要でしょうか？）
↓image_uploader.rbの中にもstrage: fogと記述している
![2017-05-04 15 33 11](https://cloud.githubusercontent.com/assets/25572309/25692744/c899b90e-30df-11e7-99d1-be60fc6869da.png)

vip-take commented on config/initializers/carrierwave.rb:
```diff
@@ -8,6 +8,12 @@
+      config.storage :fog
+    when 'test'
+      config.storage :file
+  end
```
スタイルガイドを参考にインデントを調整しましょう！
https://github.com/fortissimo1997/ruby-style-guide/blob/japanese/README.ja.md

tokuchan-pg replied:
調整いたしました！

vip-take replied:
環境ごとにバケット分けていて、Goodですね！

vip-take commented:
頂いた質問について回答します。
initializerの中に記載すれば、個別ファイルへの記載は不要になるはずです。

以下、試していないので間違っているかもしれません。

initializersに記載しているものは、グローバル設定（carriwerwave全体の設定）になり、
個別のimage_uploaderで記載することで対象のカラムだけ設定が上書きされると思います。

理由は、initializersは先に読み込まれ、その後に個別ファイルの設定が読まれるためです。

参考になるかわかりませんが、以下の記載もあります。
どちらにも設定できると書いてありますが、読まれる順番については・・どこだろう。。（今から読んでみます）
https://github.com/carrierwaveuploader/carrierwave#configuring-carrierwave

あと、さらに参考までにこのあたりを目を通しておくと良いですね！
https://railsguides.jp/initialization.html

tokuchan-pg committed:
added production env case in initializerz/carrierwave.rb

*Made two buckets in S3. One for production, the other for development.

*revised the indent style

tokuchan-pg commented:
initialierz/carrierwave.rbにproduction環境の設定を失念しておりましため、追記いたしました。（production用とdevelopment用で、S3のバケットを２つに分けました。環境変数の設定も修正済みです。）

個別ファイルへの記載の件も参考情報ありがとうございます！！チェックしてみます！！
